preprocess/docenc_main.cc

```diff
@@ -0,0 +1,256 @@
+	size_t document_index = 0;
+	std::vector<size_t>::const_iterator indices_it(indices.begin());
+
+	for (StringPiece line : in) {
+		++document_index;
+
+		if (!indices.empty()) {
+			if (*indices_it != document_index) {
+				continue; // skip document
+			} else {
+				indices_it++;
+			}
+		}
+
+		std::string document;
+		preprocess::base64_decode(line, document);
+
+		if (!delimiter_encountered && document.find(delimiter == '\n' ? std::string("\n\n") : std::string(&delimiter, 1)) != std::string::npos)
+			delimiter_encountered = true;
+
+		if (print_document_index)
+			prefix_lines(document, out, std::to_string(document_index) + "\t");
+		else
+			out << document;
+
+		out << delimiter;
+
+		// Have we found all our indices? Then stop early
+		if (!indices.empty() && indices_it == indices.end())
+			break;
+	}
+
+	return document_index;
+}
+
+size_t encode(util::FilePiece &in, util::FileStream &out, char delimiter, std::vector<size_t> const &indices) {
+	size_t document_index = 0;
+	std::string document;
+	std::vector<size_t>::const_iterator indices_it(indices.begin());
+
+	bool is_eof = false;
+	while (!is_eof) {
+		document.clear();
+		
+		// Start accumulating lines that make up a document
+		StringPiece line;
+		while (true) {
+			is_eof = !in.ReadLineOrEOF(line, delimiter, true);
+			
+			if (is_eof)
+				break;
+
+			// Is this the document delimiter when using \n\n as delimiter?
+			if (delimiter == '\n' && line.empty())
+				break;
+			
+			document.append(line.data(), line.size());
+			
+			// Add back the \n delimiter for lines
+			if (delimiter == '\n')
+				document.push_back('\n');
+
+			if (delimiter != '\n')
+				break;
+		}
+
+		// Don't bother printing anything for that last empty doc
+		if (is_eof && document.empty())
+			break;
+		
+		++document_index;
+
+		// Check whether this is a document we care about
+		if (!indices.empty()) {
+			if (*indices_it != document_index) {
+				continue; // skip document
+			} else {
+				indices_it++;
+
+				// Check whether we can stop processing altogether after this one 
+				if (indices_it == indices.end())
+					is_eof = true;
+			}
+		}
+
+		std::string encoded_document;
+		preprocess::base64_encode(StringPiece(document.data(), document.size()), encoded_document);
+		out << encoded_document << '\n';
+	}
+
+	return document_index;
+}
+
+int usage(char program_name[]) {
+	std::cerr << "Usage: "
+	     << program_name << " [ index ... ] [ files ... ]\n"
+	        "\n"
+	        "Indices:\n"
+	        "  N    Single index, starting with 1\n"
+	        "  M-N  Index range, i.e. 1-3 expands to 1 2 3\n"
+	        "\n"
+	        "Options:\n"
+	        "  -d   Decode, i.e. base64 to text (default: encode)\n"
+	        "  -0   Use nullbyte as document delimiter (default: blank line)\n"
+	        "  -q   Do not voice concerns\n"
+	        "  -v   Voice additional info\n"
```

Comment:
This is rather vague

Reply:
Oops, indeed. I've updated the descriptions (and removed the -verbose mode as it didn't really have a good use case)